File: src/io_utils.jl
module IoUtils
	using HDF5
	export Stepper, create_stepper, create_writer, write_flowfield, close_flowfield

	struct Stepper
		io_steps::Int
		total_writes::Int
	end

	function create_stepper(io_steps::Int, total_steps::Int)::Stepper
		if io_steps == 1
			total_writes = Int(floor(total_steps / io_steps))
		else
			total_writes = Int(floor(total_steps / io_steps)) + 1
		end

		return Stepper(io_steps, total_writes)
	end

	# check if we should write output now
	function should_step(stepper::Stepper, step::Int)
		return (step % stepper.io_steps == 0) || step == 1
	end

	mutable struct FlowfieldWriter
		n::Int
		h5file::HDF5.File
		u_dset::HDF5.Dataset
		p_dset::HDF5.Dataset
		t_dset::HDF5.Dataset
		ρ_dset::HDF5.Dataset
		e_dset::HDF5.Dataset
		c_dset::HDF5.Dataset
		E_dset::HDF5.Dataset
		stepper::Stepper
		write_number::Int
	end

	function create_writer(stepper::Stepper, out::T, n::Int) where T <: AbstractFloat
		filename = "./results/flowfield.h5"
		Base.Filesystem.rm(filename, force=true)

		flowfield_file_id = h5open(filename, "w")

		u_dset = create_dataset(flowfield_file_id, "velocity", datatype(T), dataspace(stepper.total_writes, n))
		p_dset= create_dataset(flowfield_file_id, "pressure", datatype(T), dataspace(stepper.total_writes, n))
		ρ_dset = create_dataset(flowfield_file_id, "rho", datatype(T), dataspace(stepper.total_writes, n))
		t_dset = create_dataset(flowfield_file_id, "time", datatype(T), dataspace(stepper.total_writes, 1))
		e_dset = create_dataset(flowfield_file_id, "entropy", datatype(T), dataspace(stepper.total_writes, n))
		c_dset = create_dataset(flowfield_file_id, "mach", datatype(T), dataspace(stepper.total_writes, n))
		E_dset = create_dataset(flowfield_file_id, "energy", datatype(T), dataspace(stepper.total_writes, n))
		
		return FlowfieldWriter(
			n,
			flowfield_file_id,
			u_dset,
			p_dset,
			t_dset,
			ρ_dset,
			e_dset,
			c_dset,
			E_dset,
			stepper,
			1
		)
	end

	function write_flowfield(
		writer::FlowfieldWriter, 
		step:: Int, 
		u::Vector{T}, 
		p::Vector{T}, 
		t::T,
		ρ::Vector{T},
		e::Vector{T},
		c::Vector{T},
		E::Vector{T}
	) where T <: AbstractFloat
		# break early if we are not supposed to step now
		if !should_step(writer.stepper, step)
			return
		end

		println("writing flowfield for step ", step, " with io steps as ", writer.stepper.io_steps, " and total writes ", writer.stepper.total_writes, "write number is ", writer.write_number)

		writer.u_dset[writer.write_number, :] = u
		writer.p_dset[writer.write_number, :] = p
		writer.t_dset[writer.write_number, 1] = t
		writer.ρ_dset[writer.write_number, :] = ρ
		writer.e_dset[writer.write_number, :] = e
		writer.c_dset[writer.write_number, :] = c
		writer.E_dset[writer.write_number, :] = E


		writer.write_number += 1
	end

	function close_flowfield(
		writer::FlowfieldWriter
	)
		close(writer.h5file)
	end
end

File: src/io_utils.rs
use crate::prelude::*;

pub(crate) struct Stepper {
    io_steps: usize,
    total_writes: usize
}

impl Stepper {
    pub(crate) fn new(io_steps: usize, total_steps: usize) -> Self {
		let total_writes = if io_steps == 1 {
			total_steps / io_steps
        } else {
			total_steps / io_steps + 1
        };

        Self { total_writes , io_steps }
    }

    fn should_step(&self, step: usize) -> bool {
        (step % self.io_steps == 0) || step == 1
    }
}

pub(crate) struct FlowfieldWriter {
    n: usize,
    stepper: Stepper,
    curr_write_number: usize,
    h5_file: hdf5::File,
    velocity: hdf5::dataset::Dataset,
    p: hdf5::dataset::Dataset,
    rho: hdf5::dataset::Dataset,
    e: hdf5::dataset::Dataset,
    c: hdf5::dataset::Dataset,
    E: hdf5::dataset::Dataset,
    t: hdf5::dataset::Dataset,
}

impl FlowfieldWriter {
    pub(crate) fn new(stepper: Stepper, n: usize) -> Self {
        let filename = "./results/flowfield.h5";
        std::fs::remove_file(filename).ok();

        let h5_file = hdf5::File::create(filename).unwrap();

        let creator = |name| {
            h5_file.new_dataset::<f64>()
                .shape((stepper.total_writes, n))
                .create(name).unwrap()
        };

        let velocity = creator("velocity");
        let p = creator("pressure");
        let rho = creator("rho");
        let e = creator("entropy");
        let c = creator("mach");
        let E = creator("energy");

        let t = h5_file.new_dataset::<f64>()
            .shape((stepper.total_writes, 1))
            .create("time").unwrap();

        FlowfieldWriter { n, stepper, curr_write_number: 0, h5_file, velocity, p, rho, e, c, E, t }
    }

    pub(crate) fn write_flowfield(&mut self, step: usize, s: &State, t: f64) {
        if !self.stepper.should_step(step) {
            return
        }

        let n = self.n;
        let write = self.curr_write_number;

        let slice = s!(write, 0..n);

        let mut mach = Vector::zeros(s.n);
        for i in 0..s.n {
            mach[i] = s.u[i] / s.c[i]
        }
        let mut time = Vector::zeros(1);
        time.fill(t);


        self.velocity.write_slice(&s.u, slice).expect("failed to write to hdf5 velocity");
        self.p.write_slice(&s.p, slice).expect("failed to write to hdf5 pressure");
        self.rho.write_slice(&s.rho, slice).expect("failed to write to hdf5 density");
        self.e.write_slice(&s.e, slice).expect("failed to write to hdf5 entropy");
        self.c.write_slice(&mach, slice).expect("failed to write to hdf5 mach");
        self.E.write_slice(&s.E, slice).expect("failed to write to hdf5 energy");
        self.t.write_slice(&time, s!(write, ..)).expect("failed to write to hdf5 time");


        self.curr_write_number += 1;
    }

    pub(crate) fn close(self) {
        self.h5_file.close().unwrap()
    }
}
